flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/autoscaler/JobVertexScaler.java

```diff
@@ -0,0 +1,264 @@
+            return false;
+        }
+
+        boolean scaleUp = currentParallelism < newParallelism;
+
+        return detectImmediateScaleDownAfterScaleUp(vertex, conf, history, scaleUp)
+                || detectIneffectiveScaleup(vertex, conf, evaluatedMetrics, history, scaleUp);
+    }
+
+    private boolean detectImmediateScaleDownAfterScaleUp(
+            JobVertexID vertex,
+            Configuration conf,
+            SortedMap<Instant, ScalingSummary> history,
+            boolean scaleUp) {
+
+        // Does not apply to scale up operations
+        if (scaleUp) {
+            return false;
+        }
+
+        var lastScalingTs = history.lastKey();
+        var lastSummary = history.get(lastScalingTs);
+
+        boolean lastScaleUp = lastSummary.getNewParallelism() > lastSummary.getCurrentParallelism();
+        if (!lastScaleUp) {
+            return false;
+        }
+
+        var gracePeriod = conf.get(SCALE_UP_GRACE_PERIOD);
+        if (lastScalingTs.plus(gracePeriod).isAfter(clock.instant())) {
+            LOG.info(
+                    "Skipping immediate scale down after scale up within grace period for {}",
+                    vertex);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private boolean detectIneffectiveScaleup(
+            JobVertexID vertex,
+            Configuration conf,
+            Map<ScalingMetric, EvaluatedScalingMetric> evaluatedMetrics,
+            SortedMap<Instant, ScalingSummary> history,
+            boolean scaleUp) {
+
+        // Does not apply to scale down operations
+        if (!scaleUp) {
+            return false;
+        }
+
+        var lastScalingTs = history.lastKey();
+        var lastSummary = history.get(lastScalingTs);
+
+        boolean lastScaleUp = lastSummary.getNewParallelism() > lastSummary.getCurrentParallelism();
+        if (!lastScaleUp) {
+            return false;
+        }
+
+        double lastProcRate = lastSummary.getMetrics().get(TRUE_PROCESSING_RATE).getAverage();
+        double lastExpectedProcRate =
+                lastSummary.getMetrics().get(EXPECTED_PROCESSING_RATE).getCurrent();
+        var currentProcRate = evaluatedMetrics.get(TRUE_PROCESSING_RATE).getAverage();
+
+        double expectedIncrease = lastExpectedProcRate / lastProcRate - 1;
+        double actualIncrease = currentProcRate / lastProcRate - 1;
+
+        boolean withinEffectiveThreshold =
+                (expectedIncrease - actualIncrease) / expectedIncrease
+                        < (1 - conf.get(AutoScalerOptions.SCALING_EFFECTIVENESS_THRESHOLD));
+
+        if (withinEffectiveThreshold) {
+            return false;
+        }
+
+        // TODO: Trigger kube event
+
+        if (conf.get(AutoScalerOptions.SCALING_EFFECTIVENESS_DETECTION_ENABLED)) {
+            LOG.info(
+                    "Skipping further scale up after ineffective previous scale up for {}", vertex);
+            return true;
```

Comment:
This condition should come earlier in the routine.

Reply:
My intention here was to trigger a kube event all the time we detect this. So even if it is disabled we would get an event. But I haven't got to actually implementing the event triggerling logic yet